core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/ShardingContainerPoolBalancer.scala

```diff
@@ -204,9 +207,33 @@ class ShardingContainerPoolBalancer(
     MetricEmitter.emitGaugeMetric(OFFLINE_INVOKER_BLACKBOX, schedulingState.blackboxInvokers.count(_.status == Offline))
   }
 
+  class RecentInvokerQueue {
+    private val maxRecentInvokers = lbConfig.maxRecentInvokers
+    private var actionLastInvokers: Queue[(Int, Long)] = Queue.empty[(Int, Long)]
+
+    def enqueue(invoker: Int): Unit = {
+      actionLastInvokers = actionLastInvokers.enqueue((invoker, Instant.now.toEpochMilli))
+      while (actionLastInvokers.size > maxRecentInvokers) { actionLastInvokers = actionLastInvokers.dequeue._2 }
+    }
+
+    /** current algorithm just orders by most frequent the last x invokers completed activations for an action */
+    def getPreferredInvokers: Seq[Int] = {
+      //filter out invokers that would represent an activation over 10 minutes old
+      val validActionLastInvokers =
+        actionLastInvokers.filter(invokerWithTimestamp => Instant.now.toEpochMilli - invokerWithTimestamp._2 < 600000)
+      if (validActionLastInvokers.size < maxRecentInvokers) {
```

Comment:
What is the reason for having this case?

Reply:
It's to bootstrap by using the home invoker algorithm. This method won't kick in in making scheduling decisions until you have the n lookback activations completed in the last ten minutes for the action.

Reply:
I'm think there should be two dials then, a min number of activations completed before making scheduling decisions with the lookback and also a max number of invokers to keep in the lookback queue.

Reply:
Yea I thought of something similar where you don't include the invoker in the final list unless the number of occurrences in the queue is over some ratio threshold. Is the concern that say the queue has three invokers in it with a max lookback of ten. Invoker 0 has 6 occurrences, invoker 1 has 3 occurrences, and invoker2 has 1 occurrence. You're think that invoker2 shouldn't be included in the returned list because the number of occurrences is too low right? I don't think that should affect things too much because the invokers it uses are still going to be bootstrapped off the home invoker algorithm and the same steps and it still prefers the most frequent as in it will try 0, then 1, then 2, and then go back to home invoker stepping. I don't think that should affect things really for an initial test